debug/src/debug.js

```diff
@@ -98,9 +116,18 @@ export function initDebug() {
 			keys.push(key);
 		}
 
-		if (oldBeforeDiff) oldBeforeDiff(vnode);
+		if (oldBeforeDiff) oldBeforeDiff(vnode, oldVnode);
 	};
 
+	// If the constructor source text and prototype properties match,
+	// two different component references are likely to be the same.
+	function componentsAreEquivalent(a, b) {
+		return (
+			Function.prototype.toString.call(a) === Function.prototype.toString.call(b) &&
```

Comment:
I did run into an issue in the past where `Function.prototype.toString` was slow (relative to Chrome) in [old versions of Firefox](https://github.com/angular/angular.js/issues/14268). I believe this has been resolved a while back, I'm just flagging it as a possible issue with this approach.

Reply:
Yeah I was iffy about this one. Also it'll break when browsers start to roll out toString censoring for functions, and JS needs the optimizations that'll unlock.

We could cache it:

```
const FNSTR = new WeakMap();
function toString(fn) {
  let str = FNSTR.get(fn);
  if (!str) {
    FNSTR.set(fn, str = Function.prototype.toString.call(fn));
  }
  return str;
}
```

Other than that, I'm not sure there's another way around this... the premise of this error is that we get a new reference to something that is roughly the same every render.

Maybe another approach would be to monitor component diffing (by WeakMap of constructors) and check if any of the children contain in-place type changes?

```
const shape = ctor => ctor.prototype && Object.keys(ctor.prototype).join();
for (let i=0; i<vnode._children.length; i++) {
  let v = vnode._children[i];
  let old = oldVnode._children[i];
  if (typeof v.type==='function' && typeof old.type==='function' && v.type!==old.type) {
    if (shape(v.type)==shape(v.type)) {
      // warn
    }
  }
}
```

Reply:
@robertknight @marvinhagemeister actually come to think of it, we could probably justify using the same caching technique for `serializeVNode()`, especially if we had it look up the source via `vnode._original` and cache based on that.

Reply:
Like this idea better in the long run, but it doesn't necessarily need to block shipping this feature as it stands.

Reply:
Shall we ship this PR as is and later do the `WeakMap` optimization?